Assertion stacktrace is not reported on `awaitItem()` timeout

### **Problem:**

Consider this test where `awaitItem()` fails due to a timeout:

```kotlin
class FlowTest {

  @Test fun fail() = runTest {
    flow {
      emit(1)
      emitAll(neverFlow()) // Avoid emitting complete
    }.test(timeout = 10.milliseconds, name = "never flow") { // L:59
      assertEquals(awaitItem(), 1)
      awaitItem() // L:61 - timeouts here
    }
  }
```

The stacktrace produced:

```java
No value produced for never flow in 10ms
java.lang.AssertionError: No value produced for never flow in 10ms
	at app.cash.turbine.ChannelKt.awaitEvent(channel.kt:92)
	at app.cash.turbine.ChannelKt$awaitEvent$1.invokeSuspend(channel.kt)
	(Coroutine boundary)
	at app.cash.turbine.FlowKt$test$2.invokeSuspend(flow.kt:57)
	at app.cash.turbine.FlowTest$fail$1.invokeSuspend(FlowTest.kt:59)
    ....
```

Note how failure trace points to `test` call on `FlowTest.kt:59`, but expected `FlowTest.kt:61`. With that, it's not obvious which of the `awaitItem()` calls timed out.

### **Cause:**

Looking at `awaitEvent()` implementation, it looks like we are rethrowing a timeout exception by creating a fresh `AssertionError` with new custom message without preserving the stacktrace:

```kotlin
/**
 * Assert that an event was received and return it.
 * This function will suspend if no events have been received.
 *
 * This function will always return a terminal event on a closed [ReceiveChannel].
 */
public suspend fun <T> ReceiveChannel<T>.awaitEvent(name: String? = null): Event<T> {
  val timeout = contextTimeout()
  return try {
    withAppropriateTimeout(timeout) {
      val item = receive()
      Event.Item(item)
    }
  } catch (e: TimeoutCancellationException) {
    throw AssertionError("No ${"value produced".qualifiedBy(name)} in $timeout")
  } catch (e: TurbineTimeoutCancellationException) {
    throw AssertionError("No ${"value produced".qualifiedBy(name)} in $timeout")
  } catch (e: CancellationException) {
    throw e
  } catch (e: ClosedReceiveChannelException) {
    Event.Complete
  } catch (e: Throwable) {
    Event.Error(e)
  }
}
```

### **(Incomplete) solution:**

If we rethrow original timeout exceptions directly...:
```kotlin
  } catch (e: TimeoutCancellationException) {
    throw e
  } catch (e: TurbineTimeoutCancellationException) {
    throw e
```

We get the stacktrace we want:
```kotlin
Timed out waiting for 10ms
app.cash.turbine.TurbineTimeoutCancellationException: Timed out waiting for 10ms
	at app//app.cash.turbine.ChannelKt$withWallclockTimeout$2$1$2.invokeSuspend(channel.kt:133)
	(Coroutine boundary)
	at app.cash.turbine.ChannelKt$withWallclockTimeout$2.invokeSuspend(channel.kt:279)
	at app.cash.turbine.ChannelKt.awaitEvent(channel.kt:85)
	at app.cash.turbine.ChannelKt.awaitItem(channel.kt:200)
	at app.cash.turbine.ChannelTurbine$awaitItem$2.invokeSuspend(Turbine.kt:191)
	at app.cash.turbine.FlowTest$fail$1$2.invokeSuspend(FlowTest.kt:61) // <--- YAY
```

This however now causes the test not to fail due to an assertion but to fail due to an unexpected error which we probably don't want.

Relevant discussion: https://androidstudygroup.slack.com/archives/C03NYGB45/p1670615690205989.

--- a/build.gradle
+++ b/build.gradle
@@ -1,3 +1,5 @@
+import org.jetbrains.kotlin.gradle.plugin.KotlinPlatformType
+
 buildscript {
   dependencies {
     classpath 'org.jetbrains.kotlin:kotlin-gradle-plugin:1.8.0'
@@ -50,6 +52,9 @@ kotlin {
         implementation 'org.jetbrains.kotlin:kotlin-test'
       }
     }
+    nonJvmTest {
+      dependsOn(commonTest)
+    }
   }
 
   sourceSets.matching { it.name.endsWith("Test") }.all {
@@ -57,6 +62,13 @@ kotlin {
       optIn('kotlinx.coroutines.DelicateCoroutinesApi')
       optIn('kotlinx.coroutines.InternalCoroutinesApi')
       optIn('kotlinx.coroutines.ExperimentalCoroutinesApi')
+    }
+  }
+
+  targets.each { target ->
+    if (target.platformType == KotlinPlatformType.common) return
+    if (target.platformType != KotlinPlatformType.jvm) {
+      target.compilations.test.defaultSourceSet.dependsOn(sourceSets.nonJvmTest)
     }
   }
 

--- a/src/commonMain/kotlin/app/cash/turbine/channel.kt
+++ b/src/commonMain/kotlin/app/cash/turbine/channel.kt
@@ -87,9 +87,9 @@ public suspend fun <T> ReceiveChannel<T>.awaitEvent(name: String? = null): Event
       Event.Item(item)
     }
   } catch (e: TimeoutCancellationException) {
-    throw AssertionError("No ${"value produced".qualifiedBy(name)} in $timeout")
+    throw TurbineAssertionError("No ${"value produced".qualifiedBy(name)} in $timeout", e)
   } catch (e: TurbineTimeoutCancellationException) {
-    throw AssertionError("No ${"value produced".qualifiedBy(name)} in $timeout")
+    throw TurbineAssertionError("No ${"value produced".qualifiedBy(name)} in $timeout", e)
   } catch (e: CancellationException) {
     throw e
   } catch (e: ClosedReceiveChannelException) {

--- a/src/commonTest/kotlin/app/cash/turbine/ChannelTest.kt
+++ b/src/commonTest/kotlin/app/cash/turbine/ChannelTest.kt
@@ -230,6 +230,11 @@ class ChannelTest {
       }
     }
     assertEquals("No value produced in 1s", actual.message)
+    assertCallSitePresentInStackTraceOnJvm(
+      throwable = actual,
+      entryPoint = "ChannelKt.awaitItem",
+      callSite = "ChannelTest\$failsOnDefaultTimeout",
+    )
   }
 
   @Test fun awaitHonorsCoroutineContextTimeoutNoTimeout() = runTest {

--- a/src/commonTest/kotlin/app/cash/turbine/FlowInScopeTest.kt
+++ b/src/commonTest/kotlin/app/cash/turbine/FlowInScopeTest.kt
@@ -165,6 +165,11 @@ class FlowInScopeTest {
       turbine.awaitItem()
     }
     assertEquals("No value produced in 1s", actual.message)
+    assertCallSitePresentInStackTraceOnJvm(
+      throwable = actual,
+      entryPoint = "ChannelTurbine\$awaitItem",
+      callSite = "FlowInScopeTest\$failsOnDefaultTimeout",
+    )
     turbine.cancel()
   }
 

--- a/src/commonTest/kotlin/app/cash/turbine/FlowTest.kt
+++ b/src/commonTest/kotlin/app/cash/turbine/FlowTest.kt
@@ -580,6 +580,11 @@ class FlowTest {
         awaitItem()
       }
       assertEquals("No value produced in 1s", actual.message)
+      assertCallSitePresentInStackTraceOnJvm(
+        throwable = actual,
+        entryPoint = "ChannelTurbine\$awaitItem",
+        callSite = "FlowTest\$failsOnDefaultTimeout",
+      )
     }
   }
 

--- a/src/commonTest/kotlin/app/cash/turbine/testUtil.common.kt
+++ b/src/commonTest/kotlin/app/cash/turbine/testUtil.common.kt
@@ -21,3 +21,15 @@ import kotlinx.coroutines.flow.flow
 
 /** A flow that never emits anything. */
 fun neverFlow(): Flow<Nothing> = flow { awaitCancellation() }
+
+/**
+ * Given a library entry point, ensure that the preceding stack frame is
+ * the expected call site.
+ *
+ * Only works on the JVM.
+ */
+expect fun assertCallSitePresentInStackTraceOnJvm(
+  throwable: Throwable,
+  entryPoint: String,
+  callSite: String,
+)

--- a/src/jvmTest/kotlin/app/cash/turbine/testUtil.kt
+++ b/src/jvmTest/kotlin/app/cash/turbine/testUtil.kt
@@ -1,0 +1,20 @@
+package app.cash.turbine
+
+actual fun assertCallSitePresentInStackTraceOnJvm(
+  throwable: Throwable,
+  entryPoint: String,
+  callSite: String,
+) {
+  val lines = throwable.stackTraceToString().lines()
+
+  val awaitItemIndex = lines.indexOfFirst { entryPoint in it }
+  if (awaitItemIndex == -1) {
+    throw AssertionError("'$entryPoint' not found in stacktrace\n\n${lines.joinToString("\n")}")
+  }
+
+  if (callSite !in lines[awaitItemIndex + 1]) {
+    throw AssertionError(
+      "Expected '$callSite' immediately precede '$entryPoint', but it did not\n\n${lines.joinToString("\n")}",
+    )
+  }
+}

--- a/src/nonJvmTest/kotlin/app/cash/turbine/testUtil.kt
+++ b/src/nonJvmTest/kotlin/app/cash/turbine/testUtil.kt
@@ -1,0 +1,9 @@
+package app.cash.turbine
+
+actual fun assertCallSitePresentInStackTraceOnJvm(
+  throwable: Throwable,
+  entryPoint: String,
+  callSite: String,
+) {
+  // Do nothing :(
+}

